File: cpp/libjoynr/capabilities/LocalDiscoveryAggregator.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2013 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include "joynr/LocalDiscoveryAggregator.h"

#include <vector>

#include "joynr/IRequestCallerDirectory.h"
#include "joynr/SystemServicesSettings.h"
#include "joynr/RequestStatus.h"
#include "joynr/RequestStatusCode.h"

#include "joynr/types/DiscoveryEntry.h"
#include "joynr/types/StdDiscoveryEntry.h"
#include "joynr/types/StdDiscoveryQos.h"
#include "joynr/system/IRouting.h"
#include "joynr/system/IDiscovery.h"
#include "joynr/system/DiscoveryProxy.h"
#include "joynr/TypeUtil.h"

namespace joynr
{

LocalDiscoveryAggregator::LocalDiscoveryAggregator(
        IRequestCallerDirectory& requestCallerDirectory,
        const SystemServicesSettings& systemServicesSettings)
        : discoveryProxy(NULL),
          hasOwnershipOfDiscoveryProxy(false),
          requestCallerDirectory(requestCallerDirectory),
          provisionedDiscoveryEntries(),
          systemServicesSettings(systemServicesSettings)
{
    QList<joynr::types::CommunicationMiddleware::Enum> connections;
    connections << joynr::types::CommunicationMiddleware::JOYNR;
    joynr::types::DiscoveryEntry routingProviderDiscoveryEntry(
            systemServicesSettings.getDomain(),
            TypeUtil::toQt(joynr::system::IRouting::INTERFACE_NAME()),
            systemServicesSettings.getCcRoutingProviderParticipantId(),
            joynr::types::ProviderQos(),
            connections);
    provisionedDiscoveryEntries.insert(
            routingProviderDiscoveryEntry.getParticipantId().toStdString(),
            routingProviderDiscoveryEntry);
    joynr::types::DiscoveryEntry discoveryProviderDiscoveryEntry(
            systemServicesSettings.getDomain(),
            TypeUtil::toQt(joynr::system::IDiscovery::INTERFACE_NAME()),
            systemServicesSettings.getCcDiscoveryProviderParticipantId(),
            joynr::types::ProviderQos(),
            connections);
    provisionedDiscoveryEntries.insert(
            discoveryProviderDiscoveryEntry.getParticipantId().toStdString(),
            discoveryProviderDiscoveryEntry);
}

LocalDiscoveryAggregator::~LocalDiscoveryAggregator()
{
    if (hasOwnershipOfDiscoveryProxy) {
        delete discoveryProxy;
    }
}

void LocalDiscoveryAggregator::setDiscoveryProxy(joynr::system::IDiscoverySync* discoveryProxy)
{
    this->discoveryProxy = discoveryProxy;
    hasOwnershipOfDiscoveryProxy = true;
}

// inherited from joynr::system::IDiscoverySync
void LocalDiscoveryAggregator::add(joynr::RequestStatus& joynrInternalStatus,
                                   const joynr::types::StdDiscoveryEntry& discoveryEntry)
{
    if (discoveryProxy == NULL) {
        joynrInternalStatus.setCode(RequestStatusCode::ERROR);
        joynrInternalStatus.addDescription(
                QString("LocalDiscoveryAggregator: discoveryProxy not set. Couldn't reach "
                        "local capabilitites directory."));
        return;
    }

    discoveryProxy->add(joynrInternalStatus, discoveryEntry);
}

void LocalDiscoveryAggregator::checkForLocalAvailabilityAndAddInProcessConnection(
        joynr::types::StdDiscoveryEntry& discoveryEntry)
{
    std::string participantId(discoveryEntry.getParticipantId());
    if (requestCallerDirectory.containsRequestCaller(participantId)) {
        std::vector<joynr::types::StdCommunicationMiddleware::Enum> connections(
                discoveryEntry.getConnections());
        connections.insert(
                connections.begin(), joynr::types::StdCommunicationMiddleware::IN_PROCESS);
        discoveryEntry.setConnections(connections);
    }
}

// inherited from joynr::system::IDiscoverySync
void LocalDiscoveryAggregator::lookup(joynr::RequestStatus& joynrInternalStatus,
                                      std::vector<joynr::types::StdDiscoveryEntry>& result,
                                      const std::string& domain,
                                      const std::string& interfaceName,
                                      const joynr::types::StdDiscoveryQos& discoveryQos)
{
    if (discoveryProxy == NULL) {
        joynrInternalStatus.setCode(RequestStatusCode::ERROR);
        joynrInternalStatus.addDescription(
                QString("LocalDiscoveryAggregator: discoveryProxy not set. Couldn't reach "
                        "local capabilitites directory."));
        return;
    }
    discoveryProxy->lookup(joynrInternalStatus, result, domain, interfaceName, discoveryQos);

    for (joynr::types::StdDiscoveryEntry& discoveryEntry : result) {
        checkForLocalAvailabilityAndAddInProcessConnection(discoveryEntry);
    }
}

// inherited from joynr::system::IDiscoverySync
void LocalDiscoveryAggregator::lookup(joynr::RequestStatus& joynrInternalStatus,
                                      joynr::types::StdDiscoveryEntry& result,
                                      const std::string& participantId)
{
    if (provisionedDiscoveryEntries.contains(participantId)) {
        joynrInternalStatus.setCode(RequestStatusCode::OK);
        result = joynr::types::DiscoveryEntry::createStd(
                provisionedDiscoveryEntries.value(participantId));
    } else {
        if (discoveryProxy == NULL) {
            joynrInternalStatus.setCode(RequestStatusCode::ERROR);
            joynrInternalStatus.addDescription(
                    QString("LocalDiscoveryAggregator: discoveryProxy not set. Couldn't reach "
                            "local capabilitites directory."));
            return;
        }
        discoveryProxy->lookup(joynrInternalStatus, result, participantId);
    }
    checkForLocalAvailabilityAndAddInProcessConnection(result);
}

// inherited from joynr::system::IDiscoverySync
void LocalDiscoveryAggregator::remove(joynr::RequestStatus& joynrInternalStatus,
                                      const std::string& participantId)
{
    if (discoveryProxy == NULL) {
        joynrInternalStatus.setCode(RequestStatusCode::ERROR);
        joynrInternalStatus.addDescription(
                QString("LocalDiscoveryAggregator: discoveryProxy not set. Couldn't reach "
                        "local capabilitites directory."));
        return;
    }
    discoveryProxy->remove(joynrInternalStatus, participantId);
}

} // namespace joynr

File: java/core/libjoynr/src/main/java/io/joynr/discovery/LocalDiscoveryAggregator.java
package io.joynr.discovery;

/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2016 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import static io.joynr.util.VersionUtil.getVersionFromAnnotation;

import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.inject.Inject;
import com.google.inject.name.Named;

import io.joynr.exceptions.JoynrRuntimeException;

import io.joynr.provider.ProviderAnnotations;
import io.joynr.proxy.Callback;
import io.joynr.proxy.Future;
import io.joynr.runtime.SystemServicesSettings;
import joynr.system.DiscoveryAsync;
import joynr.system.DiscoveryProvider;
import joynr.system.DiscoveryProxy;
import joynr.system.Routing;
import joynr.system.RoutingProvider;
import joynr.types.DiscoveryEntry;
import joynr.types.DiscoveryQos;
import joynr.types.ProviderQos;
import joynr.types.ProviderScope;

public class LocalDiscoveryAggregator implements DiscoveryAsync {

    private static final Logger logger = LoggerFactory.getLogger(LocalDiscoveryAggregator.class);

    private static final long NO_EXPIRY = Long.MAX_VALUE;
    private HashMap<String, DiscoveryEntry> provisionedDiscoveryEntries = new HashMap<>();
    private DiscoveryProxy discoveryProxy;

    @Inject
    public LocalDiscoveryAggregator(@Named(SystemServicesSettings.PROPERTY_SYSTEM_SERVICES_DOMAIN) String systemServicesDomain,
                                    @Named(SystemServicesSettings.PROPERTY_CC_DISCOVERY_PROVIDER_PARTICIPANT_ID) String discoveryProviderParticipantId,
                                    @Named(SystemServicesSettings.PROPERTY_CC_ROUTING_PROVIDER_PARTICIPANT_ID) String routingProviderParticipantId) {
        ProviderQos providerQos = new ProviderQos();
        providerQos.setScope(ProviderScope.LOCAL);
        String defaultPublicKeyId = "";
        provisionedDiscoveryEntries.put(systemServicesDomain
                + ProviderAnnotations.getInterfaceName(DiscoveryProvider.class),
                                        new DiscoveryEntry(getVersionFromAnnotation(DiscoveryProvider.class),
                                                           systemServicesDomain,
                                                           ProviderAnnotations.getInterfaceName(DiscoveryProvider.class),
                                                           discoveryProviderParticipantId,
                                                           providerQos,
                                                           System.currentTimeMillis(),
                                                           NO_EXPIRY,
                                                           defaultPublicKeyId));
        // provision routing provider to prevent lookup via discovery proxy during startup.
        provisionedDiscoveryEntries.put(systemServicesDomain + Routing.INTERFACE_NAME,
                                        new DiscoveryEntry(getVersionFromAnnotation(RoutingProvider.class),
                                                           systemServicesDomain,
                                                           Routing.INTERFACE_NAME,
                                                           routingProviderParticipantId,
                                                           providerQos,
                                                           System.currentTimeMillis(),
                                                           NO_EXPIRY,
                                                           defaultPublicKeyId));
    }

    public void setDiscoveryProxy(DiscoveryProxy discoveryProxy) {
        this.discoveryProxy = discoveryProxy;
    }

    @Override
    public Future<Void> add(Callback<Void> callback, DiscoveryEntry discoveryEntry) {
        if (discoveryProxy == null) {
            throw new JoynrRuntimeException("LocalDiscoveryAggregator: discoveryProxy not set. Couldn't reach "
                    + "local capabilitites directory.");
        }
        return discoveryProxy.add(callback, discoveryEntry);
    }

    @Override
    public Future<DiscoveryEntry[]> lookup(final Callback<DiscoveryEntry[]> callback,
                                           String[] domains,
                                           String interfaceName,
                                           DiscoveryQos discoveryQos) {
        final Set<DiscoveryEntry> discoveryEntries = new HashSet<>();
        Set<String> missingDomains = new HashSet<>();
        for (String domain : domains) {
            if (provisionedDiscoveryEntries.containsKey(domain + interfaceName)) {
                DiscoveryEntry discoveryEntry = provisionedDiscoveryEntries.get(domain + interfaceName);
                discoveryEntries.add(discoveryEntry);
            } else {
                missingDomains.add(domain);
            }
        }
        logger.trace("Found locally provisioned discovery entries: {}", discoveryEntries);
        final Future<DiscoveryEntry[]> discoveryEntryFuture = new Future<>();
        if (!missingDomains.isEmpty()) {
            logger.trace("Did not find entries for the following domains: {}", missingDomains);
            if (discoveryProxy == null) {
                throw new JoynrRuntimeException("LocalDiscoveryAggregator: discoveryProxy not set. Couldn't reach "
                        + "local capabilitites directory.");
            }
            Callback<DiscoveryEntry[]> newCallback = new Callback<DiscoveryEntry[]>() {

                @Override
                public void onFailure(JoynrRuntimeException error) {
                    callback.onFailure(error);
                    discoveryEntryFuture.onFailure(error);
                }

                @Override
                public void onSuccess(DiscoveryEntry[] entries) {
                    assert entries != null : "Entries must not be null.";
                    logger.trace("Globally found entries for missing domains: {}", entries);
                    Collections.addAll(discoveryEntries, entries);
                    resolveDiscoveryEntriesFutureWithEntries(discoveryEntryFuture, discoveryEntries, callback);
                }
            };
            String[] missingDomainsArray = new String[missingDomains.size()];
            missingDomains.toArray(missingDomainsArray);
            discoveryProxy.lookup(newCallback, missingDomainsArray, interfaceName, discoveryQos);
        } else {
            resolveDiscoveryEntriesFutureWithEntries(discoveryEntryFuture, discoveryEntries, callback);
        }
        return discoveryEntryFuture;
    }

    private void resolveDiscoveryEntriesFutureWithEntries(Future<DiscoveryEntry[]> future,
                                                          Set<DiscoveryEntry> discoveryEntries,
                                                          Callback<DiscoveryEntry[]> callback) {
        DiscoveryEntry[] discoveryEntriesArray = new DiscoveryEntry[discoveryEntries.size()];
        discoveryEntries.toArray(discoveryEntriesArray);
        future.resolve((Object) discoveryEntriesArray);
        callback.resolve((Object) discoveryEntriesArray);
    }

    @Override
    public Future<DiscoveryEntry> lookup(Callback<DiscoveryEntry> callback, String participantId) {

        if (discoveryProxy == null) {
            throw new JoynrRuntimeException("LocalDiscoveryAggregator: discoveryProxy not set. Couldn't reach "
                    + "local capabilitites directory.");
        }
        return discoveryProxy.lookup(callback, participantId);

    }

    @Override
    public Future<Void> remove(Callback<Void> callback, String participantId) {
        if (discoveryProxy == null) {
            throw new JoynrRuntimeException("LocalDiscoveryAggregator: discoveryProxy not set. Couldn't reach "
                    + "local capabilitites directory.");
        }
        return discoveryProxy.remove(callback, participantId);
    }
}
